articles/clojure_elixir_fibonacci_mstalker/article.md

```diff
@@ -0,0 +1,339 @@
+
+We could shorten our function even more by adding a [guard clause](https://hexdocs.pm/elixir/guards.html#content):
+
+```elixir
+defmodule Fibonacci do
+  def compute(n) when n in [0, 1], do: n
+  def compute(n), do: compute(n - 1) + compute(n - 2)
+end
+```
+
+A guard clause adds a condition to the function clauses. If the pattern of the argument matches the function signature, and the arguments satisfy the guard clause, the function executes; otherwise, it doesn't. In this case, I used a membership guard called [`in`](https://hexdocs.pm/elixir/Kernel.html#in/2). If `n` is equal to any element in the list `[0, 1]`, Elixir executes the first function clause and returns `n`. Otherwise, Elixir executes the second function clause.
+
+
+## Optimizing Our Functions
+
+The functions above are fine for small values of `n`. If `n` is sufficiently large, though, these functions will consume a large quantity of memory and will slow down significantly. The Clojure version even raises a `StackOverflow Error`. Try calling these functions when `n` is `200000` and you'll see what I mean!
+
+So far, our functions figured out a value by starting at a higher sequence index, and then calculating the values at lower indexes. We can speed up our function execution and lower memory consumption by starting at lower indexes and working our way up. Writing functions this way tends to be more complicated because they require one or more state variables that we didn't need in our simpler implementations above.
+
+
+### Clojure
+
+Let's take a look at a Clojure implementation.
+
+```clojure
+(ns fibonacci)
+
+(defn- helper
+  [n]
+  (loop [current-index 2
+         current-value 1
+         previous-value 1]
+    (if (= n current-index)
+      current-value
+      (recur (inc current-index)
+             (+' current-value previous-value)
+             current-value))))
+
+(defn compute
+  [n]
+  (cond (= n 0) 0
+        (= n 1) 1
+        :else (helper n)))
+```
+
+The `compute` function looks very similar to our original implementation. We return `0` or `1` if `n` is `0` or `1`, respectively. Otherwise, we execute another function to calculate the value we want. In this case we use a function called `helper`.
+
+`defn-` is a private function. Due to how Clojure compiles code, we have to define `helper` above any code that uses it.
+
+Clojure isn't tail-call-optimized, but it has a [`loop`](http://clojuredocs.org/clojure.core/loop)/[`recur`](http://clojuredocs.org/clojure.core/recur) construct that works around this. In `helper`, we set up a starting point for our recursion with `loop`, and we initialize some values. Our assumption is that if you're in this function, the index you're looking for is at least `2`, so we give `current-index` an initial value of `2`. `current-value` is `1` when the index is `2`. The previous Fibonacci value is `1`, and we call this `previous-value` in our loop.
```

Comment:
I'm not sure most readers will be familiar with tail-call optimization. Maybe add a reference?